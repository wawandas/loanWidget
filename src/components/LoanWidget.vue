<template>
  <v-container class="fill-height" fluid>
    <v-row align="center" justify="center">
      <v-col cols="8">
        <v-card
          class="elevation-4"
          tile
          :loading="isLoading"
          :disabled="isLoading"
        >
          <v-card-text>
            <v-form ref="form" v-model="valid" :lazy-validation="lazy">
              <v-container>
                <v-row>
                  <v-col cols="12" md="5">
                    <v-text-field
                      label="Amount"
                      name="amount"
                      type="text"
                      v-model="amount"
                      :rules="amountRules"
                      required
                    />
                  </v-col>
                  <v-col cols="12" md="5">
                    <v-text-field
                      label="Duration"
                      name="duration"
                      type="text"
                      v-model="duration"
                      :rules="durationRules"
                      required
                    />
                  </v-col>
                  <v-col cols="12" md="2" class="align-content-center">
                    <v-btn
                      color="primary"
                      class="mt-4"
                      block
                      @click="submit"
                      :disabled="!valid"
                    >
                      Ok
                    </v-btn>
                  </v-col>
                </v-row>
              </v-container>
            </v-form>
          </v-card-text>
        </v-card>
        <v-alert
          class="elevation-4 mt-4"
          text
          color="blue"
          v-if="monthlyInstallment"
        >
          Monthlly installment: {{ monthlyInstallment }} EUR
        </v-alert>
      </v-col>
    </v-row>
  </v-container>
</template>

<script>
import axios from "axios";

export default {
  name: "LoanWidget",

  data: () => ({
    valid: true,
    amount: "",
    amountRules: [
      v => !!v || "Amount is required",
      v => (v >= 10000 && v <= 100000) || "Should be between 10.000 and 100.000"
    ],
    duration: "",
    durationRules: [
      v => !!v || "Duration is required",
      v => (v >= 1 && v <= 5) || "Should be between 1 and 5"
    ],
    lazy: true,
    isLoading: false,
    monthlyInstallment: ""
  }),

  methods: {
    submit() {
      const isFormValid = this.$refs.form.validate();
      if (isFormValid) {
        this.getLoan();
      }
    },
    async getLoan() {
      this.isLoading = true;

      const { data } = await axios.post("/loan", {
        amount: this.amount,
        duration: this.duration
      });

      this.isLoading = false;
      this.monthlyInstallment = data.monthlyInstallment;
    }
  }
};
</script>
