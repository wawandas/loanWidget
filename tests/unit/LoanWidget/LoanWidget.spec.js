import Vue from "vue";
import Vuetify from "vuetify";
import { mount, createLocalVue } from "@vue/test-utils";
import LoanWidget from "@/components/LoanWidget.vue";

Vue.use(Vuetify);

describe("LoanWidget.vue", () => {
  let wrapper;

  beforeEach(() => {
    const localVue = createLocalVue();
    localVue.use(Vuetify);

    wrapper = mount(LoanWidget, {
      localVue
    });
  });

  it("matches snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("shows error when amount is wrong", async () => {
    const amountInput = wrapper.find('input[name="amount"]');
    amountInput.setValue("1");

    await wrapper.vm.$nextTick();

    amountInput.trigger("input");

    await wrapper.vm.$nextTick();

    expect(wrapper.text()).toContain("Should be between 10.000 and 100.000");
  });

  it("submit button is disabled when form is invalid", async () => {
    await wrapper.vm.$nextTick();

    const durationInput = wrapper.find('input[name="duration"]');
    const submitButton = wrapper.find("button");

    durationInput.setValue("7");

    await wrapper.vm.$nextTick();

    durationInput.trigger("input");

    await wrapper.vm.$nextTick();

    expect(wrapper.text()).toContain("Should be between 1 and 5");
    expect(submitButton.is("[disabled]")).toBeTruthy();
  });

  it("shows error when duration is wrong", async () => {
    const durationInput = wrapper.find('input[name="duration"]');
    durationInput.setValue("7");

    await wrapper.vm.$nextTick();

    durationInput.trigger("input");

    await wrapper.vm.$nextTick();

    expect(wrapper.text()).toContain("Should be between 1 and 5");
  });
});
