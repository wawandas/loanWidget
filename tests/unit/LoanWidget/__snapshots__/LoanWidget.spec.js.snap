// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoanWidget.vue matches snapshot 1`] = `
<div class="container fill-height container--fluid">
  <div class="row align-center justify-center">
    <div class="col col-8">
      <div class="elevation-4 v-card v-sheet v-sheet--tile theme--light">
        <div class="v-card__text">
          <form novalidate="novalidate" class="v-form">
            <div class="container">
              <div class="row">
                <div class="col-md-5 col-12">
                  <div class="v-input theme--light v-text-field">
                    <div class="v-input__control">
                      <div class="v-input__slot">
                        <div class="v-text-field__slot"><label for="input-4" class="v-label theme--light" style="right: 0px; position: absolute;">Amount</label><input name="amount" required="required" id="input-4" type="text"></div>
                      </div>
                      <div class="v-text-field__details">
                        <div class="v-messages theme--light">
                          <div class="v-messages__wrapper"></div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="col-md-5 col-12">
                  <div class="v-input theme--light v-text-field">
                    <div class="v-input__control">
                      <div class="v-input__slot">
                        <div class="v-text-field__slot"><label for="input-7" class="v-label theme--light" style="right: 0px; position: absolute;">Duration</label><input name="duration" required="required" id="input-7" type="text"></div>
                      </div>
                      <div class="v-text-field__details">
                        <div class="v-messages theme--light">
                          <div class="v-messages__wrapper"></div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div class="align-content-center col-md-2 col-12"><button type="button" class="mt-4 v-btn v-btn--block v-btn--contained theme--light v-size--default primary"><span class="v-btn__content">
                    Ok
                  </span></button></div>
              </div>
            </div>
          </form>
        </div>
      </div>
      <!---->
    </div>
  </div>
</div>
`;
